### ISSUE TITLE
Sub expressions in PSR12.ControlStructures.BooleanOperatorPlacement sniff

### ISSUE THREAD

**Issue by TRowbotham** [squizlabs/PHP_CodeSniffer#2637]
The following code currently results in 4 errors due to having multiple single line expressions in a multi-line control structure.
```php
if (
    ($n > 0 && $n < 10)
    || ($n > 10 && $n < 20)
    || ($n > 20 && $n < 30)
) {
    return $n;
}
```
```bash
------------------------------------------------------------------------------------------------------------------
FOUND 4 ERRORS AFFECTING 4 LINES
------------------------------------------------------------------------------------------------------------------
 260 | ERROR | [x] Boolean operators between conditions must be at the beginning or end of the line, but not both
 261 | ERROR | [ ] Boolean operators between conditions must be at the beginning or end of the line
 262 | ERROR | [ ] Boolean operators between conditions must be at the beginning or end of the line
 263 | ERROR | [ ] Boolean operators between conditions must be at the beginning or end of the line
```

To avoid errors, each sub expression would need to be broken up into multiple lines, such as:
```php
if (
    ($n > 0 
        && $n < 10)
    || ($n > 10 
        && $n < 20)
    || ($n > 20 
        && $n < 30)
) {
    return $n;
}
```

So, I guess the question is, if a sub expression can fit on a single line, should it be allowed or should it be forced onto multiple lines? The spec only shows a very simple expression when it comes to multi-line expressions in control structures, which leaves it open to interpretation.

**Comment by webimpress** [comment]
I got the same issue, and started investigating what PSR-12 says in that matter.

So ... PSR-12 spec says:

> Boolean operators between conditions MUST always be at the beginning or at the end of the line, not a mix of both.

https://www.php-fig.org/psr/psr-12/#51-if-elseif-else

But I am not sure how it should be interpreted. I think something like that should be disallowed:

```
if (
    $a &&
    $b
    || $c
) {
```

so we have _mixed_ - once bool operator is at the end, and the other is at the beginning.

Not sure if the wording is confusing or really it is required to have _one condition per line_.

**Comment by gsherwood** [comment]
> Not sure if the wording is confusing or really it is required to have _one condition per line_.

I have no idea.

It feels very limiting to force one condition per line. Given PSR-12 will now never be altered, I think the best thing to do might be to go against the exact wording of the standard and just ignore boolean operators that appear in the middle of lines so multiple expressions can be on the same line.

Anyone else have any thoughts on this?

**Comment by webimpress** [comment]
@gsherwood Maybe lets add configuration option for _one condition per line_?

**Comment by gsherwood** [comment]
> Maybe lets add configuration option for _one condition per line_?

We could, but we'd still need to figure out what the default value is.

**Comment by MasterOdin** [comment]
My reading of this is that for a given boolean conditional `($expr && $expr && ...)` each `$expr` is just something that resolves to true, whether it's a variable, some sort of equality, or a more complex conditional. Thus, the PSR-12 check to me would read that in each layer of the conditionals, all conditions of that layer must confirm, e.g:
```
if (
    ($n > 0 && $n < 10)
    || ($n > 10 && $n < 20)
    || ($n > 20 && $n < 30)
) {
    return $n;
}
```
is valid as the conditional operators of the outer nest are all preceeding their lines, while the inner conditionals are all on one line. This would mean would could also do:
```
if (
    (
      $expr1
      && $expr2
      && $expr3
      && $expr4
      && $expr5
      && $expr6
    )
    || ($n > 100 && $n < 200)
    || ($n > 200 && $n < 300)
) {
    return $n;
}
```

**Comment by gsherwood** [comment]
I've remove the code that produced errors when boolean operators were found in the middle of a line, so the code samples from above are now considered valid PSR12. I didn't add an option to enable this extra check because I don't really know how often it will be used, so I'll wait and see if there is support for it.

Thanks for raising this issue.
